feat(zkEVM): Add Storage SM documentation (#1293)
maticnetwork/matic-docs: docs/zkEVM/zkProver/storage-sm-design.md

---

> id: storage-sm-mechanism

Author: Yeah, it basically refers to both design and mechanism but I tried trimming the ID down to only include mechanism.

title: Storage State Machine Mechanism 
sidebar_label: Mechanism 
description: A description of Storage State Machine's mechanism.
keywords:
  - docs
  - zk rollups
  - polygon wiki
  - zkEVM
  - Polygon zkEVM
  - mechanism
image: https://wiki.polygon.technology/img/thumbnail/polygon-zkevm.png
---

> The **Storage SM** is one of zkProver's secondary state, and thus receives instructions from the Main SM called **Storage Actions**. The Storage Actions basically refers to verification of **CRUD** operations executed by the Main SM.

Author: Adding that.

> The **Storage SM** is one of zkProver's secondary state, and thus receives instructions from the Main SM called **Storage Actions**. The Storage Actions basically refers to verification of **CRUD** operations executed by the Main SM.

Author: ```suggestion
The **Storage SM** is one of zkProver's secondary state machines, and thus receives instructions from the Main SM called **Storage Actions**. The Storage Actions basically refer to the verification of **CRUD** operations executed by the Main SM.
```

The primary Assembly code uses selectors by following the sequence in which these Storage Actions are listed in the above table. That is,

- It first checks if the required action is a `Get`. If it is so, the [storage_sm_get.zkasm](https://github.com/0xPolygonHermez/zkevm-storage-rom/blob/main/zkasm/storage_sm_get.zkasm) code is fetched for execution.
- If not, it checks if the required action is `Set_Update`. If it is so, the [storage_sm_set_update.zkasm](https://github.com/0xPolygonHermez/zkevm-storage-rom/blob/main/zkasm/storage_sm_set_update.zkasm) code is fetched for execution.
- If not, it continues to check if the required action is `Set_InsertFound`. If it is so, the [storage_sm_set_insert_found.zkasm](https://github.com/0xPolygonHermez/zkevm-storage-rom/blob/main/zkasm/storage_sm_set_insert_found.zkasm) code is fetched for execution.
- If not, it continues in the same way until the correct action is selected, in which case the corresponding code is fetched for execution.   

> That's all the primary Storage Assembly code does, the details of how each if the SMT Actions are stipulated in the individual secondary Assembly codes.

Author: ```suggestion
That's all the primary Storage Assembly code does and the details of how each of the SMT Actions is stipulated in the individual secondary Assembly codes.
```

## The UPDATE zkASM Code

Take as an example the `Set_UPDATE` zkASM code. The primary Storage Assembly code uses the selector `isSetUpdate()` for `Set_UPDATE`. 

Note that an **UPDATE** action involves,

1. Reconstructs the corresponding key, from both the remaining key found at the leaf and key-bits used to navigate to the leaf.
2. Ascertains that indeed the old value was included in the old root,
3. Carries out the UPDATE of the old value with the new value, as well as updating all nodes along the path from the leaf to the root. 

There is only one `Set_UPDATE` Assembly code, [storage_sm_set_update.zkasm](https://github.com/0xPolygonHermez/zkevm-storage-rom/blob/main/zkasm/storage_sm_set_update.zkasm), for all the above three computations.

### Key Reconstruction In zkASM

Key Reconstruction is achieved in two steps; Positioning of the bit "1" in the `LEVEL` register, and using the `LEVEL` register to **climb the RKey**. That is, append the path bit last used in navigation to the correct RKey part.

1. **Positioning the bit "1" in the `LEVEL` register**

>    The `Set_UPDATE` zkASM code, first initialises the `LEVEL` register to `(1,0,0,0)`. Then uses the `GetLevelBit()` function to read the two least-significant bits of the leaf level, which happens in two cases, each with its own two subcases;

Reviewer: ```suggestion
   The `Set_UPDATE` zkASM code, first initialises the `LEVEL` register to `(1,0,0,0)`. Then uses the `GetLevelBit()` function to read the two least-significant bits of the leaf level, which happens in two cases, each with its own two subcases:
```

   The parent node is $\text{POSEIDON}\big(0\|0\|0\|0\|\text{SiblingHash}\|\text{LeafValue}\big)$. 

3. **Check if tree top has been reached**

   The code uses the function `GetTopTree()` to check is the top of the tree has been reached.

   **Case 1**. If `GetTopTree()` returns 1, then Step 2 is repeated. But this time using the hash value of the corresponding sibling at the next level (at `leaf level - 1`). 

   **Case 2**. If `GetTopTree()` returns 0, then the code jumps to the `SU_Latch` routine. 

The `SU_Latch` is an overall routine for the entire `Set_UPDATE` Assembly code. It is here where, 

- Equality between the reconstructed key and the original key is checked.

- Equality between the computed old root value and the original old root is checked.

Once consistency is established both between the keys and the old roots, then all new values; the new root, the new hash value, and the new leaf value; are set using `LATCH_SET`. 

## Remaining Secondary Assembly Codes

The Assembly codes for the other seven SMT Actions to a certain extent, follow a similar pattern except for a few cases where especially adjusted routines are used.

> Actions such as; 

Reviewer: ```suggestion
Actions such as:
```

1. The `Set_InsertFound` (or `SIF`) may involve a change in the topology of the SMT by extending a branch once or several times. 

   In cases where a branch has been extended, the SIF Assembly code, when computing the new root, uses another routine called `SIF_ClimbBranch` just for updating values along the newly extended branch. This is done in addition to the `SIF_ClimbTree`, which is the exact same routine as the aforementioned `SU_ClimbTree` of the `Set_UPDATE` case.

>    It is for the same reason, SIF Assembly utilises special registers; the `SIBLING_VALUE_HASH` and `SIBLING_RKEY`. 

Reviewer: ```suggestion
   It is for the same reason that SIF Assembly utilizes special registers: the `SIBLING_VALUE_HASH` and `SIBLING_RKEY`. 
```

2. The opposite SMT Action, the `Set_DeleteFound` or `SDF`, may entail a previously extended branch being reversed.

   As in the SIF case, if a branch had been extended but now the extension needs to be reversed due to a deleted leaf value, a special routine called `SDF_ClimbBranch` is used when updating values of nodes along the newly shortened branch. This `SDF_ClimbBranch` routine is the exact same routine as the`SIF_ClimbBranch`. Similarly, the SDF Assembly code uses the `SDF_ClimbTree` as in the Set_UPDATE Assembly.

> Note also that there is only one `Get` Assembly code, for the READ SMT Action, and the rest of the secondary Assembly codes are `Set_` Assembly codes differing according to their respective SMT Actions. So `Get` uses `LATCH_GET` at the end of a run, while the `Set_` codes use `LATCH_SET`.

Reviewer: ```suggestion
Note also that there is only one `Get` Assembly code for the READ SMT Action, and the rest of the secondary Assembly codes are `Set_` Assembly codes differing according to their respective SMT Actions. So `Get` uses `LATCH_GET` at the end of a run, while the `Set_` codes use `LATCH_SET`.
```
